bbq/compiler/compiler.go

```diff
@@ -1472,17 +1458,59 @@ func (c *Compiler[_, _]) loadTypeArguments(expression *ast.InvocationExpression)
 
 func (c *Compiler[_, _]) VisitMemberExpression(expression *ast.MemberExpression) (_ struct{}) {
 	c.compileExpression(expression.Expression)
+
+	memberAccessInfo, ok := c.ExtendedElaboration.MemberExpressionMemberAccessInfo(expression)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	if memberAccessInfo.IsOptional {
+		c.codeGen.Emit(opcode.InstructionUnwrap{})
+	}
+
 	constant := c.addStringConst(expression.Identifier.Identifier)
+
+	// TODO: remove member if `isNestedResourceMove`
+	//  See `Interpreter.memberExpressionGetterSetter` for the reference implementation.
 	c.codeGen.Emit(opcode.InstructionGetField{
 		FieldNameIndex: constant.index,
 	})
+
+	// Return a reference, if the member is accessed via a reference.
+	// This is pre-computed at the checker.
+	if memberAccessInfo.ReturnReference {
+		index := c.getOrAddType(memberAccessInfo.ResultingType)
+		c.codeGen.Emit(opcode.InstructionNewRef{
+			TypeIndex:  index,
+			IsImplicit: true,
+		})
+	}
+
+	// TODO: Need to wrap the result back with an optional, if `memberAccessInfo.IsOptional`
+
 	return
 }
 
 func (c *Compiler[_, _]) VisitIndexExpression(expression *ast.IndexExpression) (_ struct{}) {
 	c.compileExpression(expression.TargetExpression)
 	c.compileExpression(expression.IndexingExpression)
 	c.codeGen.Emit(opcode.InstructionGetIndex{})
+
+	indexExpressionTypes, ok := c.ExtendedElaboration.IndexExpressionTypes(expression)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	// Return a reference, if the element is accessed via a reference.
+	// This is pre-computed at the checker.
+	if indexExpressionTypes.ReturnReference {
```

Comment:
Same here for index expressions. In reality there are no entitlements associated in this case (no way to define mappings for array elements). It just returns a reference.